### ISSUE TITLE
Add support for Pub/Sub message ordering

### ISSUE THREAD

**Issue by meltsufin** [GoogleCloudPlatform/spring-cloud-gcp#85]
`PubSubAdmin.createSubscription()` will need to be expanded to allow enabling ordering, and whenever we create a `PubsubMessage`, we need to be able to set the ordering key for it.
Also, the Pub/Sub Cloud Stream Binder will need to be updated to support the option.
See: https://cloud.google.com/pubsub/docs/ordering
See: https://cloud.google.com/pubsub/docs/publisher#using_ordering_keys

Q: Does it makes sense to expose the keys as partitions in a Kafka-binder-like API?

**Comment by meltsufin** [comment]
@ttomsu What do you think of focusing on this one after DLQ?

**Comment by DerMorb** [comment]
The `PubSubMessageConverter.toPubSubMessage(Object payload, Map<String, String> headers)` method might also need a way to set the ordering key or we need another way to provide the ordering key.
Right now the `payload` is ending up in the `data` and the `headers` in the `attributes` of a `PubSubMessage` but the ordering key is a separate field.

Or do we need to write our own message converter somehow like this?

```
public class OrderMessagePubSubMessageConverter extends JacksonPubSubMessageConverter {

    public OrderMessagePubSubMessageConverter(ObjectMapper objectMapper) {
        super(objectMapper);
    }

    @Override
    public PubsubMessage toPubSubMessage(Object payload, Map<String, String> headers) {
        PubsubMessage pubsubMessage = super.toPubSubMessage(payload, headers);
        return PubsubMessage.newBuilder()
                            .mergeFrom(pubsubMessage)
                            .setOrderingKey(???)
                            .build();
    }
}
```

**Comment by meltsufin** [comment]
@DerMorb I think we just need to pick a name for the header by adding it to `GcpPubSubHeaders` and then extend all implementations of `PubSubMessageConverter` to marshal that header back and forth between headers and `PubsubMessage.orderingKey`. It's kind of what you're suggesting, but more general.

**Comment by meltsufin** [comment]
Now that #85 is in, what's left?

1.  In Spring Cloud Stream Binder for Pub/Sub, we should add the ability to automatically create subscriptions that support message ordering. This will involve adding `enableMessageOrdering` property to `PubSubConsumerProperties` and updating `PubSubChannelProvisioner.createSubscription`.
2. In Spring Integration Outbound Channel Adapter AKA `PubSubMessageHandler` we can add a property to specify a default ordering key.
3. Once (2) is done, we can expose this ability to the Binder by adding the `orderingKey` property to the `PubSubProducerProperties` and using it in the `PubSubMessageChannelBinder.createProducerMessageHandler`.

**Comment by vijaysdey88** [comment]
Is this feature implemented in latest release?

**Comment by meltsufin** [comment]
No, only the latest SNAPSHOT release has the change.

**Comment by pythonmeister** [comment]
Hello,

Do you have a plan when this will be published?

**Comment by elefeint** [comment]
Publisher endpoints for message ordering have been available since [version 2.0.3](https://github.com/GoogleCloudPlatform/spring-cloud-gcp/blob/main/CHANGELOG.adoc#203-2021-06-08)
